File: apps/web/src/lib/api/leaderboard.ts
import { apiGet } from "../api"

export interface LeaderboardEntry {
  rank: number
  address: string
  nickname?: string
  avatar?: string
  score: number
}

export async function fetchLeaderboard(
  type: "invite" | "bet" | "topic",
  period: "day" | "week" | "month" | "all" = "week",
): Promise<LeaderboardEntry[]> {
  const res = await apiGet(`/leaderboard/${type}?period=${period}`)
  const json = await res.json()
  return (json.data as LeaderboardEntry[]) ?? []
}

File: apps/api/app/routers/leaderboard.py
from fastapi import APIRouter, Depends, Path, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.response import success
from app.db.session import get_db
from app.services.leaderboard_service import LeaderboardService

router = APIRouter(prefix="/api/v1/leaderboard", tags=["leaderboard"])


def get_leaderboard_service(db: AsyncSession = Depends(get_db)) -> LeaderboardService:
    return LeaderboardService(db)


@router.get("/{type}")
async def get_leaderboard(
    type: str = Path(..., pattern="^(invite|bet|topic)$"),
    period: str = Query(default="week", pattern="^(day|week|month|all)$"),
    limit: int = Query(default=10, ge=1, le=100),
    svc: LeaderboardService = Depends(get_leaderboard_service),
):
    data = await svc.get_leaderboard(type, period=period, limit=limit)
    return success(data=data)
